### PR TITLE
Add better compatibility with `diffusers-interpret` (and possibly other use cases!)

### PR DESCRIPTION
**Pull request by JoaoLages**
Hi there!
I love this package ❤️

I'm the author of [diffusers-interpret](https://github.com/JoaoLages/diffusers-interpret) and along my work found these features very useful to add to this main package:
- Having the option to run `DiffusionPipeline.__call__` **while calculating gradients**;
- Having a `output_latents` flag (similar to `output_scores`/`output_attentions`/etc from `transformers`) that adds a `latents` attribute to the output;
- ~Deactivating safety checker;~ removed this option (https://github.com/huggingface/diffusers/pull/506/commits/12ca96990b0e6d7ef34a0c4bbcd9819ffe2752bd)
- Passing `text_embeddings` directly instead of the text string;
- Gradient checkpointing (this was already a feature in `transformers` too).

That's about it 😄 
To start this PR I made the changes only for the `StableDiffusionPipeline` class, but I can port those changes to the other pipelines if you agree with them.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -90,10 +90,29 @@ def disable_attention_slicing(self):
         # set slice_size = `None` to disable `attention slicing`
         self.enable_attention_slicing(None)
 
-    @torch.no_grad()
+    def gradient_checkpointing_enable(self) -> None:
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:93]: This follows the precedent set by `enable_attention_slicing` of providing methods to change this object's configuration.

But it makes for an interface where _most_ of its configuration is done through its constructor arguments, except for these few other things that can only be changed by calling these methods after its creation, which I find to be an awkward mix.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -103,13 +122,16 @@ def __call__(
         latents: Optional[torch.FloatTensor] = None,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        output_latents: bool = False,
+        run_safety_checker: bool = True,
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:126]: FYI, the team has previously shut down PRs that define an API to bypass the safety filter, e.g. #254.

I recommend leaving that out of this PR just so that doesn't get mixed up with the other features you propose here.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -161,33 +190,52 @@ def __call__(
...
+        elif torch.is_tensor(prompt):
+            if len(prompt.shape) == 2:
+                # Add batch dimension
+                prompt = prompt.unsqueeze(0)
+
+            if len(prompt.shape) != 3:
+                raise ValueError(
+                    f"If `prompt` is of type `torch.Tensor`, it is expected to have a 2 dimensions "
+                    f"(sequence len, embedding dim) or 3 dimensions (batch size, sequence len, embedding dim), "
+                    f"but found tensor with shape {prompt.shape}"
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:210]: +1 for checking the input shape here. Can you add a note about that to its docstring too?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -274,6 +327,9 @@ def __call__(
             image = self.numpy_to_pil(image)
 
         if not return_dict:
-            return (image, has_nsfw_concept)
+            return (image, has_nsfw_concept, all_latents)
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:330]: I think if PipelineOutput classes are the way forward and the tuple return format here is mainly for backwards compatibility, we should leave it the same size it was (a pair) and not worry about adding new features to it.

**Review by keturn — COMMENTED** [review]
Hi João! I was just introduced to `diffusers-interpret` yesterday via [the discord](https://discord.gg/G7tWnz98XR)! I have all the same questions so I love seeing that sort of thing.

I have no authority to merge anything here, but I've taken the liberty of leaving a few notes.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 1 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:93]: Agreed, at least I would make it a constructor argument as well, but I tried to make it similar to what we have in the `transformers` package.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 2 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:126]: Sure, I can remove that part if needed 👍

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 4 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:330]: Since this method had a `@torch.no_grad` decorator, I don't think this is for backwards compatibility 🤔 
But looking at the transformers package, it seems that[ when `return_dict_generate=False`, options like `output_scores`/`output_attentions` don't matter](https://github.com/huggingface/transformers/blob/main/src/transformers/generation_utils.py#L2078), so it makes sense to remove `latents` from the tuple as you mention :)

File: src/diffusers/pipelines/stable_diffusion/__init__.py [hunk 5]
```diff
@@ -21,10 +21,13 @@ class StableDiffusionPipelineOutput(BaseOutput):
         nsfw_content_detected (`List[bool]`)
             List of flags denoting whether the corresponding generated image likely represents "not-safe-for-work"
             (nsfw) content.
+        latents (`List[torch.Tensor]`, *optional*, returned when `output_latents=True` and `return_dict=True` is passed)
+            List (one element for each diffusion step) of `torch.Tensor` of shape `(batch_size, in_channels, height // 8, width // 8)`
     """
 
     images: Union[List[PIL.Image.Image], np.ndarray]
     nsfw_content_detected: List[bool]
+    latents: Optional[List[torch.Tensor]] = None
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/__init__.py:30]: Fine with me! What do you think @pcuenca @anton-l @patil-suraj ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 6]
```diff
@@ -103,14 +122,18 @@ def __call__(
...
+        enable_grad: bool = False,
         **kwargs,
     ):
         r"""
         Function invoked when calling the pipeline for generation.
 
         Args:
-            prompt (`str` or `List[str]`):
-                The prompt or prompts to guide the image generation.
+            prompt (`str`, `List[str]` or `torch.Tensor`):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:134]: Ufff, I don't think the input `prompt` should ever be a tensor that's confusing and opens the box for hacky code - can't we just work with the `latents` inputs?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 7]
```diff
@@ -161,33 +191,52 @@ def __call__(
...
 
+        # enable/disable grad
+        was_grad_enabled = torch.is_grad_enabled()
+        torch.set_grad_enabled(enable_grad)
+
         if isinstance(prompt, str):
             batch_size = 1
         elif isinstance(prompt, list):
             batch_size = len(prompt)
+        elif torch.is_tensor(prompt):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:202]: I don't think `prompt` should ever be a tensor

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 8]
```diff
@@ -259,6 +309,10 @@ def __call__(
             else:
                 latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
 
+            if output_latents:
+                # save latents from all diffusion steps
+                all_latents.append(latents)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:314]: fine with this! Think this makes a lot of sense

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 9]
```diff
@@ -161,33 +191,52 @@ def __call__(
                 device = "cuda" if torch.cuda.is_available() else "cpu"
             self.to(device)
 
+        # enable/disable grad
+        was_grad_enabled = torch.is_grad_enabled()
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:195]: This will also be a bit difficult to accept for me. It's a) a bit hacky to me and b) Pipelines by definition should only be used for inference. I assume the gradients are needed for analysis and the idea is not to do training? 
I'm not 100% sure whether they are enough use cases that warrant allowing gradient flow here on the other hand it also shouldn't hurt really if we leave the default to `False`. IMO working with function decorators and `enable_grad` + `disable_grad` functions  is the way to go here though instead.

What do you think @patil-suraj @anton-l @pcuenca ?

**Review by patrickvonplaten — COMMENTED** [review]
Hey @JoaoLages,

Super cool interpret library btw! Like the idea of returning the latents and could also be convinced to allow gradient computation (even though I think the use case is too niche for now and would only be pro if all the community really wants this feature cc @hysts @pcuenca @patil-suraj @anton-l - wdyt?). Don't think we should allow the prompt to be a torch.Tensor.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:134]: What if we use `inputs_embeds` as some methods in `transformers` have?

Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:134]: This is what I meant: https://github.com/huggingface/transformers/blob/16913b3c9215f592f1240511f8da271dc07c3552/src/transformers/models/t5/modeling_t5.py#L905

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 2 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py]: removed in https://github.com/huggingface/diffusers/pull/506/commits/12ca96990b0e6d7ef34a0c4bbcd9819ffe2752bd

File: src/diffusers/pipelines/stable_diffusion/__init__.py (hunk 5 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/__init__.py:30]: fine with me as well.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 1 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:93]: gradient checkpointing will be integrated soon #461

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:133]: Agree with Patrick.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 9 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:192]: Same as Patrick, not really in favor of this.

> IMO working with function decorators and enable_grad + disable_grad functions is the way to go here though instead.

+1

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR @JoaoLages !

I have pretty much same comments as Patrick.

- We could definitely return intermidiate latents
- gradient checkpointing will be supported soon. It should not be included in pipeline like this. Pipelines are intended for inference only so it's best to avoid training related logic here.
- enabling gradients: We could add this if community is really interested in it. Could you please open an issue for this ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:133]: The problem is that we can pass an image to StableDiffusionImg2ImgPipeline.__call__ as a tensor but we can't pass a text... 
Would and extra `inputs_embeds` or `prompt_embeds` argument work as an alternative?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 1 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:93]: Awesome! will remove from this PR 👍

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 9 above)
Comment (JoaoLages) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:192]: > IMO working with function decorators and enable_grad + disable_grad functions is the way to go here though instead.

By that means, keeping `@torch.no_grad` decorator in `__call__` and add put the whole method under `with torch.enable_grad() if enable_grad else nullcontext():` ?
Not sure what you meant.

**Comment by JoaoLages** [comment]
> * gradient checkpointing will be supported soon. It should not be included in pipeline like this. Pipelines are intended for inference only so it's best to avoid training related logic here.

🚀 

> * enabling gradients: We could add this if community is really interested in it. Could you please open an issue for this ?

There you go https://github.com/huggingface/diffusers/issues/529

**Comment by patrickvonplaten** [comment]
Closing this PR as it does too many changes at once -> happy to continue the discussion on the single PRs that were opened :-)
